src/snap.c

```diff
@@ -0,0 +1,284 @@
+		.right  = view->pending.x + view->pending.width  + margin.right,
+		.bottom = view->pending.y + view->pending.height + margin.bottom
+	};
+	return edge;
+}
+
+struct border
+snap_get_max_distance(struct view *view)
+{
+	struct output *output = view->output;
+	struct border margin = ssd_get_margin(view->ssd);
+	struct wlr_box usable = output_usable_area_scaled(output);
+	struct border distance = {
+		.left   = usable.x + margin.left + rc.gap - view->pending.x,
+		.top    = usable.y + margin.top  + rc.gap - view->pending.y,
+		.right  = usable.x + usable.width  - view->pending.width
+			- margin.right  - rc.gap - view->pending.x,
+		.bottom = usable.y + usable.height - view->pending.height
+			- margin.bottom - rc.gap - view->pending.y
+	};
+	return distance;
+}
+
+struct snap_search {
+	const int search_dir; /* -1: left/up, 1: right/down */
+
+	const int add_view_x;
+	const int add_view_y;
+	const int add_view_width;
+	const int add_view_height;
+
+	const int add_margin_left;
+	const int add_margin_top;
+	const int add_margin_right;
+	const int add_margin_bottom;
+};
+
+/* near/far is the left, right, top or bottom border of a window,
+ * depending on the search direction:
+ *  - near_right: search to the right, snap to left  (near) border of a window.
+ *  - far_right:  search to the right, snap to right (far)  border of a window.
+ *  - near_left:  search to the left,  snap to right (near) border of a window.
+ *  - far_left:   search to the left,  snap to left  (far)  border of a window.
+ *
+ * structs below define what coordinates and margins to take into
+ * account depending near/far, and direction.
+ */
+static const struct snap_search near_left  = { -1,   1, 0, 1, 0,    0,  0,  1,  0 };
+static const struct snap_search near_up    = { -1,   0, 1, 0, 1,    0,  0,  0,  1 };
+static const struct snap_search near_right = {  1,   1, 0, 0, 0,   -1,  0,  0,  0 };
+static const struct snap_search near_down  = {  1,   0, 1, 0, 0,    0, -1,  0,  0 };
+static const struct snap_search far_left   = { -1,   1, 0, 0, 0,   -1,  0,  0,  0 };
+static const struct snap_search far_up     = { -1,   0, 1, 0, 0,    0, -1,  0,  0 };
+static const struct snap_search far_right  = {  1,   1, 0, 1, 0,    0,  0,  1,  0 };
+static const struct snap_search far_down   = {  1,   0, 1, 0, 1,    0,  0,  1,  0 };
+
+static inline int
+_snap_next_edge(struct view *view, int start_pos, const struct snap_search def, int max, int gap)
+{
+	struct output *output = view->output;
+	struct server *server = output->server;
+	struct view *v;
+	int p = max;
+	struct wlr_scene_tree *current_workspace_tree = view->server->workspace_current->tree;
+	wl_list_for_each(v, &server->views, link) {
+		if (v == view || v->output != output || v->minimized || v->maximized) {
+			continue;
+		}
+		/* not part of the current workspace, and not part of the always-on-top tree */
+		if (v->scene_tree->node.parent != current_workspace_tree &&
+			!view_is_always_on_top(v)) {
+			continue;
+		}
```

Comment:
Checking workspaces, as suggested @Consolatis here: https://github.com/labwc/labwc/pull/1005#discussion_r1281295788

Works for me, please double-check.

Reply:
Yep, seems to work fine. We should provide a helper function like `view_is_in_viewport()` or something along that line later on but we don't have to do that in this PR.